### PR TITLE
Modbus transformation support

### PR DESCRIPTION
**Pull request by ssalonen**
This is a pull request introducing transformation support to the modbus binding.

In addition, a new (backwards compatible) extension is introduced to the item configuration, allowing the following
- read-only items
- write-only items
- valueType overriden per-item basis to something else than the slave setting
- processing only certain states ("trigger")
- processing only certain commands ("trigger")
- having multiple "inbound" and "outbound" IO rules describing conversion of modbus data to openhab, and vice versa (enables complex use cases such as roller shutter items and custom serialization logic, see below)

After this all item types are supported, as long as user provides the binding the necessary rules to deserialize/serialize the state/commands. Documentation provides several practical examples, e.g. one way of communicating roller shutter item commands via modbus.

Noteworthy items
- the change is significant in the sense that even "old style" item configurations are parsed to quite different data structure (`ItemIOConnection`), and these new data structures are used in the binding read/write functions. However, there is pretty good regression test set covering those
- INCREASE/DECREASE are handled slightly differently. Previously `valueType` (e.g. `int32`) was ignored; currently it is used.
- code uses java8 language features such as lambdas and streams. I have understood this should be now ok
- the previosu roller shutter example removed from the documentation examples in favour of the new example; should we still keep it as long as 1.9.0 is the latest release version?
- copyright notice changed when running the maven command. Is it OK now?
- the actual helper class conducting the transformation (`Transformation`) heavily influenced copy-pasted from other bindings (http or mqtt, don't remember anymore)

Community post discussing the feature: https://community.openhab.org/t/modbus-support-for-transformations-roller-shutter-items-read-only-items-write-only-items-and-others/21500

Unfortunately not too much user feedback, but some
- [custom serialization logic @ community forums](https://community.openhab.org/t/solved-modbus-and-cheap-4-relay-module-d18994/21711/5?u=ssalonen) -- solved user issue
- positive feedback on transformation support: 
  - [Sprocky @ community forums](https://community.openhab.org/t/modbus-support-for-transformations-roller-shutter-items-read-only-items-write-only-items-and-others/21500/8?u=ssalonen) 
  - [cerberek @ community](https://community.openhab.org/t/modbus-support-for-transformations-roller-shutter-items-read-only-items-write-only-items-and-others/21500/6?u=ssalonen)
  - [@paulbeech @ #4204](https://github.com/openhab/openhab1-addons/issues/4204#issuecomment-277280464): transformation used to scale modbus values on read successfully
- [@primary1 @ #4745](https://github.com/openhab/openhab1-addons/issues/4745#issuecomment-281945090): resolves the issue (tested on openHAB1)

Solves #4204 (feature request for transformation support)
Solves #4745 (functionality regression introduced while back)
Solves #4407 
Relieves pain points of #4654 (roller shutter items as an example now easier to handle)

### CODE REVIEW

**Comment by ssalonen** [comment]
Hi

A little over month passed now... Anything I can do to move this forward?

Best,
Sami

**Comment by 9037568** [comment]
Kinda thought John was taking this one, since he put the labels on it, but I can start a review...

**Comment by 9037568** [comment]
I haven't been able to get the tests to compile here. Any chance there are some classpath changes you need to commit?

**Comment by 9037568** [comment]
````
[ERROR] import org.openhab.binding.modbus.internal.pooling.ModbusSerialSlaveEndpoint;
[ERROR] ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[ERROR] The import org.openhab.binding.modbus.internal.pooling cannot be resolved
[ERROR] \openhab\bundles\binding\org.openhab.binding.modbus.test\src\test\java\org\openhab\binding\modbus\internal\ModbusSlaveEndpointTestCase.java:[14]
[ERROR] import org.openhab.binding.modbus.internal.pooling.ModbusTCPSlaveEndpoint;
[ERROR] ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[ERROR] The import org.openhab.binding.modbus.internal.pooling cannot be resolved
````

**Comment by ssalonen** [comment]
Is that from eclipse IDE?

Should not require any class path changes. Could you try to mvn install the modbus binding.

Sami

**Comment by ssalonen** [comment]
Checked also the [Cloudbees CI](https://openhab.ci.cloudbees.com/job/openHAB1-Addons/org.openhab.binding$org.openhab.binding.modbus.test/) -- it seems to run the tests successfully.

There should be no classpath changes or anything like this change set :o

**Comment by 9037568** [comment]
This was from mvn. I did an mvn clean install of modbus first, which ran fine, then the same for modbus.test, which produced the errors shown above.

And I think that Cloudbees will not be running with your changes, yet so...

But I'll keep poking at it...

**Comment by ssalonen** [comment]
This is what I did

1. `mvn install`  in `modbus`
2. `mvn verify` in `modbus.test`. Log displays that locally installed version is used:

```
[INFO] Resolving dependencies of MavenProject: org.openhab.binding:org.openhab.binding.modbus.test:1.10.0-SNAPSHOT @ /home/salski/src/openhab2-master/git/openhab/bundles/binding/org.openhab.binding.modbus.test/pom.xml
[WARNING] The following locally built units have been used to resolve project dependencies:
[WARNING]   org.openhab.binding.modbus/1.10.0.201703241650
[INFO] Resolving class path of MavenProject: org.openhab.binding:org.openhab.binding.modbus.test:1.10.0-SNAPSHOT @ /home/salski/src/openhab2-master/git/openhab/bundles/binding/org.openhab.binding.modbus.test/pom.xml
```

and all tests are run...
```
Results :

Tests run: 748, Failures: 0, Errors: 0, Skipped: 0

[INFO] All tests passed!

....

[INFO] BUILD SUCCESS
```



`mvn install` works the same way for me.

Do you have that in your output when running `mvn install?`

I wonder what I have different?


Best,
Sami

**Comment by 9037568** [comment]
I'm still unable to build the tests, but I'm going to ignore that. Thanks for this contribution!

File: bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/SimpleTokenizer.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+import org.apache.commons.lang.StringEscapeUtils;
+
+/**
+ * String tokenizer for splitting strings. Understands quoting and escaped quotes.
+ *
+ * See tests for details
+ *
+ * @author Sami Salonen
+ *
```
Comment (ssalonen) [inline bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/SimpleTokenizer.java]: Done

File: bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusBindingConfig.java [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+        } else {
+            // Number items
+            return UnDefType.UNDEF;
+        }
+    }
+
+    /**
+     * Constructor for config string
+     *
+     * Config can needs to of the following basic formats
```
Comment (ssalonen) [inline bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusBindingConfig.java]: Changed the wording, hopefully more clear now.

File: bundles/binding/org.openhab.binding.modbus.test/src/test/java/org/openhab/binding/modbus/internal/ModbusBindingConfigTest.java [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+                        new Object[] { new StringItem("item2"),
+                                ">[slave1:99:trigger=*,transformation=JS(getValue.js),valueType=int32]\t  ,   >[slave2:98:trigger=*,transformation=\"REGEX(\\\",.=*[:]),)\",valueType=uint16]",
+                                new ItemIOConnection[0],
+                                new ItemIOConnection[] {
+                                        new ItemIOConnection("slave1", 99, IOType.COMMAND, "*",
+                                                new Transformation("", "JS", "getValue.js"), "int32"),
+                                        new ItemIOConnection("slave2", 98, IOType.COMMAND, "*",
+                                                new Transformation("", "REGEX", "\",.=*[:]),"), "uint16") },
+                                null },
+                        // // extended, unexpected "type" (commented as type not supported as of now)
```
Comment (ssalonen) [inline bundles/binding/org.openhab.binding.modbus.test/src/test/java/org/openhab/binding/modbus/internal/ModbusBindingConfigTest.java]: Done

File: bundles/binding/org.openhab.binding.modbus.test/src/test/java/org/openhab/binding/modbus/internal/ModbusBindingConfigTest.java [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+                                                + "'invalidValueType' does not match expected: "
+                                                + "'bit, int8, uint8, int16, uint16, int32, uint32, float32, int32_swap, "
+                                                + "uint32_swap, float32_swap or 'default'") },
+                        new Object[] { new StringItem("item2"), "<[slave1:99:foobarKey=*]", null, null,
+                                new BindingConfigParseException(
+                                        "Parsing of item 'item2' configuration '<[slave1:99:foobarKey=*]' (as part of the whole "
+                                                + "config '<[slave1:99:foobarKey=*]') failed: "
+                                                + "org.openhab.model.item.binding.BindingConfigParseException Unexpected token "
+                                                + "'foobarKey=*, expecting key to be one of: type, trigger, transformation, valueType") },
+                        // // default type (commented as type not supported as of now)
```
Comment (ssalonen) [inline bundles/binding/org.openhab.binding.modbus.test/src/test/java/org/openhab/binding/modbus/internal/ModbusBindingConfigTest.java]: Done

File: bundles/binding/org.openhab.binding.modbus.test/src/test/java/org/openhab/binding/modbus/internal/ConfigUpdatedTestCase.java [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2010-2016, openHAB.org and others.
+ * Copyright (c) 2010-2016 by the respective copyright holders.
```
Comment (ssalonen) [inline bundles/binding/org.openhab.binding.modbus.test/src/test/java/org/openhab/binding/modbus/internal/ConfigUpdatedTestCase.java]: Done

File: bundles/binding/org.openhab.binding.modbus/README.md [hunk 6]
```diff
@@ -119,55 +116,110 @@ Examples:
...
-Contact Contact1 "Contact1 [MAP(en.map):%s]" (All)   {modbus="slave2:0"}
-```
+In addition to simple format, a new configuration syntax was introduced in version 1.10.0.
+
+This gives more freedom to the user. Among other things it enables
+- specify transformation to use on read/write
+- mark item as read-only (received openHAB commands are not processed)
+- mark item as write-only (polled data do not change the item state)
+- overriding `valueType` per-item basis (allows to read many registers at once and then interpret them using different value types)
+- processing only certain command (on write)
```
Comment (ssalonen) [inline bundles/binding/org.openhab.binding.modbus/README.md]: Done

File: bundles/binding/org.openhab.binding.modbus/README.md [hunk 7]
```diff
@@ -282,9 +414,10 @@ When the binding interprets and converts polled input registers (`input`) or hol
 
 - 1. register(s) are first parsed to a number (see below for the details, exact logic depends on `valuetype`)
 - 2a. if the item is Switch or Contact: zero is converted CLOSED / OFF. Other numbers are converted to OPEN / ON.
-- 2b. if the item is Number: the value is passed as is
+- 2b. if the item is Number: the value is used as is
+- 3. transformation is done to the value, if configured. The transformation output (string) is parsed to state using item's accepted state types (e.g. number, or CLOSED/OPEN).
 
-The logic for converting read registers to number goes as below. Different procedure is taken depending on `valuetype`. 
+The logic for converting to the read read registers to number goes as below. Different procedure is taken depending on `valuetype`.
```
Comment (ssalonen) [inline bundles/binding/org.openhab.binding.modbus/README.md]: Changed the wording, hopefully more clear now.

File: bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ItemIOConnection.java [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+import org.apache.commons.lang.builder.ToStringBuilder;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.State;
+
+/**
+ * ItemIOConnection defines the translation of data from modbus to openhab, and vice versa.
+ *
+ *
+ * @author Sami Salonen
+ *
```
Comment (ssalonen) [inline bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ItemIOConnection.java]: Done

File: bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusGenericBindingProvider.java [hunk 9]
```diff
@@ -66,17 +61,18 @@ public String getBindingType() {
...
+        } else if (item.getClass() == ContactItem.class) {
             return;
-        }
-        if (item.getClass() == NumberItem.class) {
+        } else if (item.getClass() == NumberItem.class) {
             return;
+        } else {
+            logger.debug(
+                    "Item '{}' is of type '{}'. Please make sure that transformation is "
+                            + "inplace to convert to polled data to a format understood "
```
Comment (ssalonen) [inline bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/ModbusGenericBindingProvider.java]: Now
```
             logger.debug(
                    "Item '{}' is of type '{}'. Please make sure that transformation is "
                            + "in place to convert the polled data to a format understood "
                            + "by the item. Furthermore, make sure that commands are converted to "
                            + "DecimalType or any command accepted by Switch, Contact or Number items.",
                    item.getName(), item.getClass().getSimpleName());
```

File: bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/Transformation.java [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Class describing transformation of a command or state.
+ *
+ * Inspired from other openHAB binding "Transformation" classes.
+ *
+ * @author Sami Salonen
+ *
```
Comment (ssalonen) [inline bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/Transformation.java]: Done

File: bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/Transformation.java [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+        public TransformationService getTransformationService(BundleContext context, String transformationServiceName) {
+            return TransformationHelper.getTransformationService(context, transformationServiceName);
+        }
+    }
+
+    public static final String TRANSFORM_DEFAULT = "default";
+    public static final Transformation IDENTITY_TRANSFORMATION = new Transformation(TRANSFORM_DEFAULT, null, null);
+    private static final TransformationHelperWrapper DEFAULT_TRANSFORMATION_HELPER = new TransformationHelperWrapperImpl();
+
+    /** RegEx to extract a parse a function String <code>'(.*?)\((.*)\)'</code> */
```
Comment (ssalonen) [inline bundles/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/Transformation.java]: Done

**Comment by ssalonen** [comment]
Thank you for taking time to do the review. I added additional commits resolving the issues you found.

**Comment by 9037568** [comment]
Looks like an Eclipse format run is still needed...

**Comment by ssalonen** [comment]
@9037568 Thanks for reminding me, forgot that initially, sorry!

Not the `modbus` and `modbus.test` has been run via ESH source code cleanup rules.

Best,
Sami

**Comment by 9037568** [comment]
Thanks for all your work on this, @ssalonen !

**Pull request merged**